### PR TITLE
Implement the API endpoint for listing releases

### PR DESCRIPTION
**Pull request by 417-72KI**
Implements the [API endpoint to list releases](https://developer.github.com/v3/repos/releases/#list-releases) on a GitHub repository.

Also fixes some properties which will be `null` when the draft release is returned from API.

### CODE REVIEW

File: OctoKit/Releases.swift [hunk 1]
```diff
@@ -49,6 +49,25 @@ public struct Release: Codable {
...
+    /// Fetches the list of releases.
+    /// - Parameters:
+    ///   - session: RequestKitURLSession, defaults to URLSession.shared()
+    ///   - owner: The user or organization that owns the repositories.
+    ///   - repository: The name of the repository.
+    ///   - completion: Callback for the outcome of the fetch.
+    @discardableResult
+    func listReleases(_ session: RequestKitURLSession = URLSession.shared, owner: String, repository: String, completion: @escaping (_ response: Response<[Release]>) -> Void) -> URLSessionDataTaskProtocol? {
+        let router = ReleaseRouter.listReleases(configuration, owner, repository)
+        return router.load(session, dateDecodingStrategy: .formatted(Time.rfc3339DateFormatter), expectedResultType: [Release].self) { releases, error in
```
Comment (MaxDesiatov) [inline OctoKit/Releases.swift:61]: Diffs aren't automatically wrapped everywhere, so please avoid long lines, as it makes the code harder to review on devices without a wide screen. 100-120 characters line limit should be sufficient in most cases.

File: Tests/OctoKitTests/ReleasesTests.swift [hunk 2]
```diff
@@ -11,12 +11,53 @@ import OctoKit
...
     static var allTests = [
+        ("testListReleases", testListReleases),
         ("testPostRelease", testPostRelease),
         ("testLinuxTestSuiteIncludesAllTests", testLinuxTestSuiteIncludesAllTests)
     ]
 
     // MARK: Actual Request tests
 
+    func testListReleases() {
+        let session = OctoKitURLTestSession(expectedURL: "https://api.github.com/repos/octocat/Hello-World/releases", expectedHTTPMethod: "GET", jsonFile: "Fixtures/releases", statusCode: 200)
```
Comment (MaxDesiatov) [inline Tests/OctoKitTests/ReleasesTests.swift:22]: *nitpick*: please split this to multiple lines to reduce the length of this line for easier review.

**Comment by 417-72KI** [comment]
@MaxDesiatov 
Thank you for reviewing!

I'm ready to split long lines, but if I do that, it will be different style from other lines.
Is it acceptable?

**Comment by MaxDesiatov** [comment]
Maybe some formatting rules applied with SwiftFormat could be introduced in a separate PR? I'll defer to @AvdLee on this matter

File: OctoKit/Releases.swift (hunk 1 above)
Comment (pietbrauer) [reply OctoKit/Releases.swift:61]: I agree that we should limit the lines to some amount of characters. I don’t know if 100-120 is the correct number but it should be enforced with some swiftformat/swiftlint rules so it applies everywhere for everyone.

I personally wouldn’t leave this as a mandatory change to get this PR accepted.

File: OctoKit/Releases.swift [hunk 3]
```diff
@@ -18,8 +18,8 @@ public struct Release: Codable {
     public let url: URL
     public let htmlURL: URL
     public let assetsURL: URL
-    public let tarballURL: URL
-    public let zipballURL: URL
+    public let tarballURL: URL?
+    public let zipballURL: URL?
```
Comment (pietbrauer) [inline OctoKit/Releases.swift:22]: Can you elaborate why this change was necessary? It would be a breaking change.

File: OctoKit/Releases.swift [hunk 4]
```diff
@@ -28,7 +28,7 @@ public struct Release: Codable {
     public let draft: Bool
     public let prerelease: Bool
     public let createdAt: Date
-    public let publishedAt: Date
+    public let publishedAt: Date?
```
Comment (pietbrauer) [inline OctoKit/Releases.swift:31]: Can you elaborate why this change was necessary? It would be a breaking change.

File: OctoKit/Releases.swift (hunk 3 above)
Comment (417-72KI) [reply OctoKit/Releases.swift:22]: These properties return null and cause `DecodingError` when `draft` is true.
Tarball and zipball are not available on draft release.

File: OctoKit/Releases.swift (hunk 4 above)
Comment (417-72KI) [reply OctoKit/Releases.swift:31]: Same reason as above.

File: Tests/OctoKitTests/ReleasesTests.swift [hunk 5]
```diff
@@ -11,12 +11,53 @@ import OctoKit
...
+                    XCTFail("Failed to unwrap `releases.first`")
+                }
+                if let release = releases.last {
+                    XCTAssertEqual(release.tagName, "v1.0.0")
+                    XCTAssertEqual(release.commitish, "master")
+                    XCTAssertEqual(release.name, "v1.0.0 Release")
+                    XCTAssertEqual(release.body, "The changelog of this release")
+                    XCTAssertFalse(release.prerelease)
+                    XCTAssertFalse(release.draft)
+                    XCTAssertNotNil(release.tarballURL)
```
Comment (pietbrauer) [inline Tests/OctoKitTests/ReleasesTests.swift:47]: I just noticed, could we assert the correct URL here as you did with the Strings? Else this looks fine to me and we can merge it.

File: OctoKit/Releases.swift (hunk 1 above)
Comment (AvdLee) [reply OctoKit/Releases.swift:61]: @MaxDesiatov this should do the job: https://realm.github.io/SwiftLint/line_length.html

Comment (MaxDesiatov) [reply OctoKit/Releases.swift:61]: I'm aware of this one, and SwiftFormat recently also gained an ability to fit code into given line length too 🙂

File: Tests/OctoKitTests/ReleasesTests.swift (hunk 5 above)
Comment (417-72KI) [reply Tests/OctoKitTests/ReleasesTests.swift]: Yes we could, and I fixed with using `XCTAssertEqual`.

**Review by pietbrauer — APPROVED** [review]

**Comment by pietbrauer** [comment]
Thanks a lot!
